test/jdk/javax/swing/JButton/SwingButtonResizeTestWithOpenGL.java

```diff
@@ -0,0 +1,306 @@
+/*
+ * Copyright (c) 2015, 2024, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+import javax.imageio.ImageIO;
+import javax.swing.JButton;
+import javax.swing.JFrame;
+import javax.swing.SwingUtilities;
+import java.awt.Color;
+import java.awt.Dimension;
+import java.awt.FlowLayout;
+import java.awt.Point;
+import java.awt.Rectangle;
+import java.awt.Robot;
+import java.awt.event.FocusAdapter;
+import java.awt.event.FocusEvent;
+import java.awt.image.BufferedImage;
+import java.awt.image.ColorModel;
+import java.awt.image.Raster;
+import java.io.File;
+import java.io.IOException;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicReference;
+
+/*
+ * @test
+ * @bug 8338103
+ * @key headful
+ * @summary Verifies that the OpenGL pipeline does not create artifacts
+ * with swing components after window is zoomed to maximum size and then
+ * resized back to normal.  The test case simulates this operation using
+ * a JButton.  A file image of the component will be saved before and after
+ * window resize. The test passes if both button images are the same.
+ * @run main/othervm -Dsun.java2d.opengl=True -Dsun.java2d.opengl.fbobject=false SwingButtonResizeTestWithOpenGL
```

Comment:
Does this really work well on Windows ?  Commonly OpenGL tests fail on Windows.
Did you test it there ?

Also is there any reason we can't test this with additional pipelines on specific platforms ?
The behaviours being tested are Java behaviours so it ought to pass in other cases too.

Reply:
Yes, I tested it on different windows platforms (Windows 10,11,2022) and it succeeded everywhere.

Reply:
@prrace I’m not sure I understand what you meant by 'testing this with additional pipelines on specific platforms.' Are you suggesting that we should test this not only with -Dsun.java2d.opengl=True -Dsun.java2d.opengl.fbobject=false, but also with other parameters?

Reply:
Pipeline. You can't easily know what will be chosen on a test system, so you need to do things like
@run main/othervm -Dsun.java2d.xrender=false
@run main/othervm -Dsun.java2d.xrender=true
@run main/othervm -Dsun.java2d.d3d=false
@run main/othervm -Dsun.java2d.d3d=true

In some cases, nothing will change but if the test is cheap and fast, nothing lost

Reply:
Thanks @prrace , I have added more pipelines to the test, please have a look. I have rerun this again on all platforms with mach5 and the results are updated in the bug as a comment.

Reply:
Also, I see you added metal.
I think I'm going to request an extra layer on top of what I previously suggested.

It is possible to have multiple @test tags, not just @run tags, so you can do platform-specific things on just the specific platform, so it would become

 /* @test 
    @requires("os.name"="windows")
   @run main/othervm -Dsun.java2d.d3d=false
   @run main/othervm -Dsun.java2d.d3d=true
*/

 /* @test 
    @requires("os.name"="macos")
   @run main/othervm -Dsun.java2d.d3d=false
   @run main/othervm -Dsun.java2d.metal=true
*/


 /* @test 
    @requires("os.name"="linux")
   @run main/othervm -Dsun.java2d.xrender=false
   @run main/othervm -Dsun.java2d.xrender=true
*/

and finally the one that is cross-platform AND defaults
 /* @test 
   @run/main/othervm
   @run main/othervm -Dsun.java2d.opengl=false
   @run main/othervm -Dsun.java2d.opengl=true
*/

Reply:
> Looks like it failed on Linux with the default pipeline, which should be xrender. That's a bit surprising. You should verify this and if necessary file a bug and exclude that failing combination from testing.

Yea, that's true, but it failed on a particular machine only and that too it was a plain run without any flags(@run main/othervm SwingButtonResizeTestWithOpenGL), the runs with xrender and other flags were successful in the same machine. I just rerun it 10 times on different linux machines and I couldn't observe the failure again, so it looks like a machine specific issue and I could see some UI debris from the previous runs on the screenshots. anyway I will check it again with @test tags.

Reply:
@prrace I have added @test tags for different platforms and the corresponding mach5 runs are attached in the bug as a new comment, please have a look.